File: triplate-java/src/test/java/dev/triplate/ApiTest.java
package dev.triplate;

import dev.triplate.Ast.BindingKeySym;
import dev.triplate.Ast.CommentSym;
import dev.triplate.Ast.IriSym;
import dev.triplate.Ast.LiteralSym;
import dev.triplate.Ast.LoopDeclSym;
import dev.triplate.Ast.LoopRefSym;
import dev.triplate.Ast.ParamDeclSym;
import dev.triplate.Ast.ParamRefSym;
import dev.triplate.Ast.PnameSym;
import dev.triplate.Ast.TemplateSymbol;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class ApiTest {

  /** Build a minimal template: a frontmatter header with the given declarations plus a body. */
  private static String h(String decls, String body) {
    return "---\nparams { " + decls + " }\n---\n" + body;
  }

  @Test
  void frontmatterIsStripped() {
    String out = Triplate.render(h("s: iri", "?x a ${s}"), Map.of("s", "http://example.org/C"));
    assertEquals("?x a <http://example.org/C>", out);
  }

  @Test
  void missingHeaderThrowsSyntaxError() {
    assertThrows(TriplateSyntaxError.class, () -> Triplate.render("SELECT * WHERE { ?s ?p ?o }", Map.of()));
  }

  @Test
  void undeclaredVariableThrowsAtCompileTime() {
    assertThrows(TriplateSyntaxError.class, () -> Triplate.compile(h("s: iri", "?x a ${t}")));
  }

  @Test
  void frontmatterPrefixesFromExamplesAndLiteralTypes() {
    CompiledTemplate tmpl = Triplate.compile(
        "---\n"
            + "params {\n  type: iri\n  amount: literal(xsd:decimal)\n  note: string\n}\n"
            + "example demo \"Demo\" {\n  type: schema:Person\n  amount: \"5\"\n  note: \"n\"^^my:dt\n}\n"
            + "---\n${type}");
    assertEquals(Set.of("my", "schema", "xsd"), tmpl.frontmatterPrefixes());
  }

  @Test
  void frontmatterPrefixesIgnoresFullIri() {
    CompiledTemplate tmpl = Triplate.compile(
        "---\nparams {\n  type: iri\n}\nexample demo \"D\" {\n  type: <http://example.org/Person>\n}\n---\n${type}");
    assertTrue(tmpl.frontmatterPrefixes().isEmpty());
  }

  @Test
  void unknownParameterThrowsBindingError() {
    assertThrows(TriplateBindingError.class,
        () -> Triplate.render(h("s: iri", "?x a ${s}"),
            Map.of("s", "http://example.org/C", "extra", "nope")));
  }

  @Test
  void iriTemplatePercentEncodesHoles() {
    String out = Triplate.render(
        h("id: string", "?s ex:p $<http://example.org/person/${id}>"),
        Map.of("id", "a b/c"));
    assertEquals("?s ex:p <http://example.org/person/a%20b%2Fc>", out);
  }

  @Test
  void rawHoleBreakingIriIsRejected() {
    assertThrows(TriplateTypeError.class,
        () -> Triplate.render(h("x: raw", "?s a $<http://example.org/${x}>"),
            Map.of("x", "a> . ?s ?p ?o . <b")));
  }

  @Test
  void forLoopWithExplicitJoinIsVerbatim() {
    String out = Triplate.render(
        h("xs: int[]", "{% for x in xs join \",\" explicit %}${x}{% endfor %}"),
        Map.of("xs", List.of(1, 2, 3)));
    assertEquals("1,2,3", out);
  }

  @Test
  void cardinalityBelowMinThrows() {
    assertThrows(TriplateCardinalityError.class,
        () -> Triplate.render(h("xs: iri[] min 1", "${xs}"), Map.of("xs", List.of())));
  }

  @Test
  void termTypeSerializesNamedNode() {
    Term namedNode = new Term() {
      @Override public String termType() { return "NamedNode"; }
      @Override public String value() { return "http://example.org/x"; }
    };
    String out = Triplate.render(h("t: term", "?s a ${t}"), Map.of("t", namedNode));
    assertEquals("?s a <http://example.org/x>", out);
  }

  @Test
  void previewExampleRendersExampleSet() {
    String template = """
        ---
        params { service: iri, limit: int }
        example demo "DBpedia" {
          service: <http://dbpedia.org/sparql>
          limit: 10
        }
        ---
        SERVICE ${service} { ?s ?p ?o } LIMIT ${limit}""";
    CompiledTemplate t = Triplate.compile(template);
    assertEquals(1, t.examples().size());
    String out = t.previewExample("demo");
    assertEquals("SERVICE <http://dbpedia.org/sparql> { ?s ?p ?o } LIMIT 10", out);
  }

  @Test
  void doubleUsesCanonicalScientificNotation() {
    assertEquals("FILTER(?x > 1.5E6)",
        Triplate.render(h("d: double", "FILTER(?x > ${d})"), Map.of("d", 1500000)));
    assertTrue(Triplate.render(h("d: double", "${d}"), Map.of("d", 0)).equals("0.0E0"));
  }

  private static final String SYM_SRC =
      "---\n"
          + "params {\n  who: pname\n  amount: literal(xsd:decimal)\n}\n"
          + "example demo \"D\" {\n  who: schema:Person\n  amount: \"9.99\"^^xsd:decimal\n  home: <http://ex.org/me>\n}\n"
          + "---\n"
          + "?s a ${who} . {% if amount %}${amount}{% endif %}";

  private static List<String> names(List<TemplateSymbol> syms, Class<? extends TemplateSymbol> kind) {
    List<String> out = new ArrayList<>();
    for (TemplateSymbol s : syms) {
      if (kind.isInstance(s)) {
        if (s instanceof ParamDeclSym d) out.add(d.name());
        else if (s instanceof ParamRefSym r) out.add(r.name());
        else if (s instanceof BindingKeySym b) out.add(b.name());
      }
    }
    return out;
  }

  @Test
  void symbolSpansSliceTheirOwnText() {
    for (TemplateSymbol s : Triplate.compile(SYM_SRC).symbols()) {
      String slice = SYM_SRC.substring(s.start(), s.end());
      String expected;
      if (s instanceof PnameSym p) expected = p.prefix() + ":" + p.local();
      else if (s instanceof IriSym i) expected = "<" + i.value() + ">";
      else if (s instanceof LiteralSym l) expected = "\"" + l.value() + "\"";
      else if (s instanceof ParamDeclSym d) expected = d.name();
      else if (s instanceof ParamRefSym r) expected = r.name();
      else if (s instanceof BindingKeySym b) expected = b.name();
      else throw new AssertionError("unexpected symbol: " + s);
      assertEquals(expected, slice);
    }
  }

  @Test
  void symbolsCaptureEveryKind() {
    List<TemplateSymbol> syms = Triplate.compile(SYM_SRC).symbols();
    assertEquals(List.of("who", "amount"), names(syms, ParamDeclSym.class));
    // Two `amount` refs: the {% if amount %} condition and the ${amount} hole.
    assertEquals(List.of("who", "amount", "amount"), names(syms, ParamRefSym.class));
    assertEquals(List.of("who", "amount", "home"), names(syms, BindingKeySym.class));
    // pname: the literal(xsd:decimal) type, the schema:Person value, the ^^xsd:decimal datatype.
    assertEquals(3, syms.stream().filter(s -> s instanceof PnameSym).count());
    List<TemplateSymbol> iris = syms.stream().filter(s -> s instanceof IriSym).toList();
    assertEquals(1, iris.size());
    assertEquals("http://ex.org/me", ((IriSym) iris.get(0)).value());
    List<TemplateSymbol> lits = syms.stream().filter(s -> s instanceof LiteralSym).toList();
    assertEquals(1, lits.size());
    assertEquals("9.99", ((LiteralSym) lits.get(0)).value());
    assertEquals("xsd:decimal", ((LiteralSym) lits.get(0)).datatype());
  }

  @Test
  void symbolsAreInAscendingSourceOrder() {
    int prev = -1;
    for (TemplateSymbol s : Triplate.compile(SYM_SRC).symbols()) {
      assertTrue(s.start() >= prev, "symbols must be in ascending source order");
      prev = s.start();
    }
  }

  @Test
  void hashIsPlainTextInBodyNotAComment() {
    CompiledTemplate tmpl = Triplate.compile("---\nparams { title: raw }\n---\n# ${title}");
    assertEquals("# My Title", tmpl.render(Map.of("title", "My Title")));
  }

  @Test
  void frontmatterCommentsArePositionedSymbols() {
    String src =
        "---\n"
            + "# a standalone comment\n"
            + "params {\n  who: pname  # the subject\n}\n"
            + "---\n"
            + "?s a ${who}";
    List<CommentSym> comments =
        Triplate.compile(src).symbols().stream()
            .filter(s -> s instanceof CommentSym)
            .map(s -> (CommentSym) s)
            .toList();
    assertEquals(
        List.of("# a standalone comment", "# the subject"),
        comments.stream().map(CommentSym::value).toList());
    for (CommentSym c : comments) {
      assertEquals(c.value(), src.substring(c.start(), c.end()));
    }
    // Body comments stay text, not symbols — comments are frontmatter-only.
    assertTrue(
        Triplate.compile("---\nparams { x: int }\n---\n# body ${x}").symbols().stream()
            .noneMatch(s -> s instanceof CommentSym));
    // Comments do not disturb ascending source order.
    int prev = -1;
    for (TemplateSymbol s : Triplate.compile(src).symbols()) {
      assertTrue(s.start() >= prev, "symbols must be in ascending source order");
      prev = s.start();
    }
  }

  @Test
  void nestedFrontmatterCommentsArePositionedSymbols() {
    String src =
        "---\n"
            + "# a header note\n"
            + "params {\n"
            + "  u: {\n"
            + "    id:   iri     # in a record type\n"
            + "    name: string\n"
            + "  }\n"
            + "  tags: string[]  # in params\n"
            + "}\n"
            + "example demo {\n"
            + "  u: { id: <http://example.org/1>,   # in an example record\n"
            + "       name: \"Alice\" }\n"
            + "  tags: [\n"
            + "    \"a\",          # in an example list\n"
            + "    \"b\"\n"
            + "  ]\n"
            + "}\n"
            + "---\n"
            + "${u.id} ${...tags}";
    List<TemplateSymbol> symbols = Triplate.compile(src).symbols();
    List<CommentSym> comments =
        symbols.stream().filter(s -> s instanceof CommentSym).map(s -> (CommentSym) s).toList();
    // A comment is legal wherever an item may start, at every nesting depth.
    assertEquals(
        List.of(
            "# a header note",
            "# in a record type",
            "# in params",
            "# in an example record",
            "# in an example list"),
        comments.stream().map(CommentSym::value).toList());
    for (CommentSym c : comments) {
      assertEquals(c.value(), src.substring(c.start(), c.end()));
    }
    int prev = -1;
    for (TemplateSymbol s : symbols) {
      assertTrue(s.start() >= prev, "symbols must be in ascending source order");
      prev = s.start();
    }
  }

  @Test
  void moduleSymbolsIsLenientOnMalformedTemplate() {
    String bad = "---\nparams { a: int }\nexample x {\n  who: schema:Person\n";
    assertThrows(TriplateSyntaxError.class, () -> Triplate.compile(bad));
    List<TemplateSymbol> syms = Triplate.symbols(bad);
    assertEquals(3, syms.size());
    assertTrue(syms.get(0) instanceof ParamDeclSym);
    assertTrue(syms.get(1) instanceof BindingKeySym);
    assertTrue(syms.get(2) instanceof PnameSym);
  }

  private static List<LoopDeclSym> loopDecls(List<TemplateSymbol> syms) {
    return syms.stream().filter(s -> s instanceof LoopDeclSym).map(s -> (LoopDeclSym) s).toList();
  }

  private static List<LoopRefSym> loopRefs(List<TemplateSymbol> syms) {
    return syms.stream().filter(s -> s instanceof LoopRefSym).map(s -> (LoopRefSym) s).toList();
  }

  @Test
  void forLoopEmitsLoopDeclAndLoopRefsInOneScope() {
    String src = "---\nparams { graphIris: pname[] }\n---\n{% for g in graphIris %} FROM ${g} ${g} {% endfor %}";
    List<TemplateSymbol> syms = Triplate.symbols(src);
    // The for source is an ordinary parameter reference, not a loop ref.
    assertEquals(
        List.of("graphIris"),
        syms.stream().filter(s -> s instanceof ParamRefSym).map(s -> ((ParamRefSym) s).name()).toList());
    List<LoopDeclSym> decls = loopDecls(syms);
    List<LoopRefSym> refs = loopRefs(syms);
    assertEquals(List.of("g"), decls.stream().map(LoopDeclSym::name).toList());
    assertEquals(List.of("g", "g"), refs.stream().map(LoopRefSym::name).toList());
    // Declaration and both references share the one scope id.
    Set<Integer> ids = new java.util.HashSet<>();
    ids.add(decls.get(0).scope());
    refs.forEach(r -> ids.add(r.scope()));
    assertEquals(1, ids.size());
    // Spans still ascend (the item precedes the source in the header).
    int prev = -1;
    for (TemplateSymbol s : syms) {
      assertTrue(s.start() >= prev, "symbols must be in ascending source order");
      prev = s.start();
    }
  }

  @Test
  void loopSymbolSpansSliceTheirOwnText() {
    String src = "---\nparams { xs: pname[] }\n---\n{% for g in xs %}${g}{% endfor %}";
    for (TemplateSymbol s : Triplate.symbols(src)) {
      if (s instanceof LoopDeclSym d) assertEquals(d.name(), src.substring(d.start(), d.end()));
      else if (s instanceof LoopRefSym r) assertEquals(r.name(), src.substring(r.start(), r.end()));
    }
  }

  @Test
  void loopShadowingBindsRefsToTheirOwnScope() {
    String src =
        "---\nparams { a: pname[]\n  b: pname[] }\n---\n"
            + "{% for g in a %}${g}{% for g in b %}${g}{% endfor %}${g}{% endfor %}";
    List<TemplateSymbol> syms = Triplate.symbols(src);
    List<LoopDeclSym> decls = loopDecls(syms);
    assertEquals(2, decls.size());
    int outer = decls.get(0).scope();
    int inner = decls.get(1).scope();
    assertTrue(outer != inner);
    // Refs in source order: outer ${g}, inner ${g}, outer ${g}.
    assertEquals(
        List.of(outer, inner, outer), loopRefs(syms).stream().map(LoopRefSym::scope).toList());
  }

  @Test
  void loopVariableShadowsSameNamedParameter() {
    String src = "---\nparams { g: pname\n  xs: pname[] }\n---\n${g} {% for g in xs %}${g}{% endfor %} ${g}";
    List<TemplateSymbol> syms = Triplate.symbols(src);
    // Two ${g} outside the loop stay paramRefs; the in-loop ${g} is a loopRef.
    assertEquals(
        List.of("g", "xs", "g"),
        syms.stream().filter(s -> s instanceof ParamRefSym).map(s -> ((ParamRefSym) s).name()).toList());
    List<LoopRefSym> refs = loopRefs(syms);
    assertEquals(List.of("g"), refs.stream().map(LoopRefSym::name).toList());
    assertEquals(loopDecls(syms).get(0).scope(), refs.get(0).scope());
  }

  @Test
  void dottedLoopReferenceMarksOnlyTheRoot() {
    String src = "---\nparams { xs: pname[] }\n---\n{% for g in xs %}${g.field}{% endfor %}";
    assertEquals(List.of("g"), loopRefs(Triplate.symbols(src)).stream().map(LoopRefSym::name).toList());
  }

  @Test
  void moduleSymbolsCollectsLoopSymbolsOnMalformedTemplate() {
    String bad = "---\nparams { xs: pname[] }\n---\n{% for g in xs %}${g} {% if";
    List<TemplateSymbol> syms = Triplate.symbols(bad);
    assertTrue(syms.stream().anyMatch(s -> s instanceof LoopDeclSym));
    assertTrue(syms.stream().anyMatch(s -> s instanceof LoopRefSym));
  }
}

File: triplate-ts/test/api.test.ts
import { describe, expect, it } from 'vitest';
import {
  compile,
  registerType,
  render,
  symbols,
  TriplateBindingError,
  TriplateSyntaxError,
  TriplateTypeError,
} from '../src/index.js';

const H = (decls: string, body: string) => `---\nparams { ${decls} }\n---\n${body}`;

describe('API behaviour', () => {
  it('compile once, render many', () => {
    const tmpl = compile(H('c: iri', '?s a ${c}'));
    expect(tmpl.render({ c: 'http://example.org/A' })).toContain('<http://example.org/A>');
    expect(tmpl.render({ c: 'http://example.org/B' })).toContain('<http://example.org/B>');
  });

  it('exposes the declared schema', () => {
    const tmpl = compile(H('s: iri, n: int optional', '${s}'));
    expect(tmpl.schema.params.map((p) => p.name)).toEqual(['s', 'n']);
    expect(tmpl.schema.byName.n.optional).toBe(true);
  });

  it('frontmatter is stripped — no leakage before the body', () => {
    const tmpl = compile('---\nparams {\n  c: iri\n}\n\n---\nSELECT * WHERE { ?s a ${c} }');
    expect(tmpl.render({ c: 'http://example.org/A' })).toBe('SELECT * WHERE { ?s a <http://example.org/A> }');
  });

  it('previewExample renders a named example set', () => {
    const tmpl = compile(
      '---\n' +
        'params {\n  classes: iri[]\n}\n' +
        'example demo "Demo" {\n  classes: [ ex:Person, <http://example.org/Org> ]\n}\n' +
        '---\n' +
        'PREFIX ex: <http://example.org/>\n' +
        'SELECT * WHERE {\n{% for c in classes join "UNION" %}\n  { ?s a ${c} }\n{% endfor %}\n}',
    );
    expect(tmpl.examples.map((e) => e.id)).toEqual(['demo']);
    const out = tmpl.previewExample('demo');
    expect(out).toContain('{ ?s a <http://example.org/Person> } UNION { ?s a <http://example.org/Org> }');
  });

  it('bigint serializes as int', () => {
    expect(render(H('x: int', '${x}'), { x: 9007199254740993n })).toBe('9007199254740993');
  });

  it('Date objects serialize for dateTime', () => {
    const d = new Date('2024-03-01T12:00:00.000Z');
    expect(render(H('x: dateTime', '${x}'), { x: d })).toBe(
      '"2024-03-01T12:00:00.000Z"^^<http://www.w3.org/2001/XMLSchema#dateTime>',
    );
  });

  it('RDF/JS terms serialize via the term type', () => {
    const named = { termType: 'NamedNode', value: 'http://example.org/x' };
    expect(render(H('t: term', '${t}'), { t: named })).toBe('<http://example.org/x>');
    const bad = { termType: 'NamedNode', value: 'http://x/> . } DROP ALL #' };
    expect(() => render(H('t: term', '${t}'), { t: bad })).toThrow(TriplateTypeError);
  });

  it('custom types can be registered (extensibility)', () => {
    registerType('uuidref', (value, pos) => {
      if (typeof value !== 'string' || !/^[0-9a-f-]{36}$/.test(value)) {
        throw new TriplateTypeError('invalid uuid', pos.line, pos.column);
      }
      return `<urn:uuid:${value}>`;
    });
    expect(render(H('id: uuidref', '${id}'), { id: '123e4567-e89b-12d3-a456-426614174000' })).toBe(
      '<urn:uuid:123e4567-e89b-12d3-a456-426614174000>',
    );
  });

  it('IRI templates percent-encode holes and validate the result', () => {
    expect(render(H('id: string', '$<http://ex.org/${id}>'), { id: 'a/b é' })).toBe(
      '<http://ex.org/a%2Fb%20%C3%A9>',
    );
    expect(() => render(H('x: raw', '$<http://ex.org/${x}>'), { x: 'a> <b' })).toThrow(TriplateTypeError);
  });

  it('explicit join is verbatim, default join pads', () => {
    const body = '{% for c in xs join "," %}${c}{% endfor %}';
    expect(render(H('xs: string[]', body), { xs: ['a', 'b'] })).toBe('"a" , "b"');
    const ebody = '{% for c in xs join "," explicit %}${c}{% endfor %}';
    expect(render(H('xs: string[]', ebody), { xs: ['a', 'b'] })).toBe('"a","b"');
  });

  it('frontmatterPrefixes recovers prefixes from example values and literal types', () => {
    const tmpl = compile(
      '---\n' +
        'params {\n  type: iri\n  amount: literal(xsd:decimal)\n  note: string\n}\n' +
        'example demo "Demo" {\n  type: schema:Person\n  amount: "5"\n  note: "n"^^my:dt\n}\n' +
        '---\n${type}',
    );
    expect([...tmpl.frontmatterPrefixes()].sort()).toEqual(['my', 'schema', 'xsd']);
  });

  it('frontmatterPrefixes ignores full <iri> values (no prefix)', () => {
    const tmpl = compile(
      '---\nparams {\n  type: iri\n}\nexample demo "D" {\n  type: <http://example.org/Person>\n}\n---\n${type}',
    );
    expect(tmpl.frontmatterPrefixes().size).toBe(0);
  });

  it('undeclared variable is a compile-time error', () => {
    expect(() => compile(H('s: iri', '${t}'))).toThrow(TriplateSyntaxError);
  });

  it('missing required parameter throws at render', () => {
    expect(() => render(H('s: iri', '${s}'), {})).toThrow(TriplateBindingError);
  });

  it('a template without frontmatter is rejected', () => {
    expect(() => compile('SELECT * WHERE { ?s ?p ?o }')).toThrow(TriplateSyntaxError);
  });

  describe('positioned symbols', () => {
    const SRC =
      '---\n' +
      'params {\n  who: pname\n  amount: literal(xsd:decimal)\n}\n' +
      'example demo "D" {\n  who: schema:Person\n  amount: "9.99"^^xsd:decimal\n  home: <http://ex.org/me>\n}\n' +
      '---\n' +
      '?s a ${who} . {% if amount %}${amount}{% endif %}';

    it('every symbol span slices its own text from the source', () => {
      for (const s of compile(SRC).symbols()) {
        const slice = SRC.slice(s.start, s.end);
        const expected =
          s.kind === 'pname'
            ? `${s.prefix}:${s.local}`
            : s.kind === 'iri'
              ? `<${s.value}>`
              : s.kind === 'literal'
                ? JSON.stringify(s.value)
                : s.name;
        expect(slice).toBe(expected);
      }
    });

    it('captures decls, refs, binding keys, pnames, iris and literals', () => {
      const byKind = (k: string) => compile(SRC).symbols().filter((s) => s.kind === k);
      expect(byKind('paramDecl').map((s) => (s as { name: string }).name)).toEqual(['who', 'amount']);
      // Two `amount` refs: the {% if amount %} condition and the ${amount} hole.
      expect(byKind('paramRef').map((s) => (s as { name: string }).name)).toEqual(['who', 'amount', 'amount']);
      expect(byKind('bindingKey').map((s) => (s as { name: string }).name)).toEqual(['who', 'amount', 'home']);
      // pname: the literal(xsd:decimal) type, the schema:Person value, and the ^^xsd:decimal datatype.
      expect(byKind('pname')).toHaveLength(3);
      expect(byKind('iri').map((s) => (s as { value: string }).value)).toEqual(['http://ex.org/me']);
      const lits = byKind('literal') as Array<{ value: string; datatype?: string }>;
      expect(lits).toEqual([{ kind: 'literal', value: '9.99', datatype: 'xsd:decimal', start: expect.any(Number), end: expect.any(Number) }]);
    });

    it('symbols are emitted in ascending source order', () => {
      const offsets = compile(SRC).symbols().map((s) => s.start);
      expect(offsets).toEqual([...offsets].sort((a, b) => a - b));
    });

    it('grouping paramDecl + paramRef + bindingKey by name yields every rename site', () => {
      const sites = compile(SRC)
        .symbols()
        .filter((s) => (s.kind === 'paramDecl' || s.kind === 'paramRef' || s.kind === 'bindingKey') && (s as { name: string }).name === 'amount');
      expect(sites.map((s) => s.kind)).toEqual(['paramDecl', 'bindingKey', 'paramRef', 'paramRef']);
    });

    it('`#` is plain text in the body, not a comment — directives after it are live', () => {
      const tmpl = compile('---\nparams { title: raw }\n---\n# ${title}');
      expect(tmpl.render({ title: 'My Title' })).toBe('# My Title');
    });

    it('`#` in the frontmatter is a syntax error, not a comment', () => {
      const src = '---\nparams { who: pname }\n# a comment\n---\n?s a ${who}';
      expect(() => compile(src)).toThrow(TriplateSyntaxError);
    });

    it('the standalone symbols() is lenient — returns what parsed on a malformed template', () => {
      const bad = '---\nparams { a: int }\nexample x {\n  who: schema:Person\n';
      expect(() => compile(bad)).toThrow(TriplateSyntaxError);
      expect(symbols(bad).map((s) => s.kind)).toEqual(['paramDecl', 'bindingKey', 'pname']);
    });

    describe('loop variables', () => {
      it('a for loop emits a loopDecl at the item plus a loopRef per body ref, all one scope', () => {
        const src = '---\nparams { graphIris: pname[] }\n---\n{% for g in graphIris %} FROM ${g} ${g} {% endfor %}';
        const syms = symbols(src);
        // The for source is an ordinary parameter reference, not a loop ref.
        const srcRef = syms.find((s) => s.kind === 'paramRef') as { name: string } | undefined;
        expect(srcRef?.name).toBe('graphIris');
        const decls = syms.filter((s) => s.kind === 'loopDecl') as Array<{ name: string; scope: number }>;
        const refs = syms.filter((s) => s.kind === 'loopRef') as Array<{ name: string; scope: number }>;
        expect(decls.map((d) => d.name)).toEqual(['g']);
        expect(refs.map((r) => r.name)).toEqual(['g', 'g']);
        // Declaration and both references share the one scope id.
        expect(new Set([decls[0].scope, ...refs.map((r) => r.scope)]).size).toBe(1);
        // Spans still ascend (the item precedes the source in the header).
        const offsets = syms.map((s) => s.start);
        expect(offsets).toEqual([...offsets].sort((a, b) => a - b));
      });

      it('every loop symbol span slices its own text', () => {
        const src = '---\nparams { xs: pname[] }\n---\n{% for g in xs %}${g}{% endfor %}';
        for (const s of symbols(src)) {
          if (s.kind === 'loopDecl' || s.kind === 'loopRef') {
            expect(src.slice(s.start, s.end)).toBe(s.name);
          }
        }
      });

      it('shadowing — inner refs bind to the inner scope, outer refs to the outer', () => {
        const src = '---\nparams { a: pname[]\n  b: pname[] }\n---\n{% for g in a %}${g}{% for g in b %}${g}{% endfor %}${g}{% endfor %}';
        const syms = symbols(src);
        const decls = syms.filter((s) => s.kind === 'loopDecl') as Array<{ scope: number }>;
        const refs = syms.filter((s) => s.kind === 'loopRef') as Array<{ scope: number }>;
        expect(decls).toHaveLength(2);
        const [outer, inner] = [decls[0].scope, decls[1].scope];
        expect(outer).not.toBe(inner);
        // Refs in source order: outer ${g}, inner ${g}, outer ${g}.
        expect(refs.map((r) => r.scope)).toEqual([outer, inner, outer]);
      });

      it('a loop variable shadowing a same-named parameter — only in-scope refs become loopRef', () => {
        const src = '---\nparams { g: pname\n  xs: pname[] }\n---\n${g} {% for g in xs %}${g}{% endfor %} ${g}';
        const syms = symbols(src);
        // Two ${g} outside the loop stay paramRefs; the in-loop ${g} is a loopRef.
        expect((syms.filter((s) => s.kind === 'paramRef') as Array<{ name: string }>).map((s) => s.name)).toEqual(['g', 'xs', 'g']);
        const refs = syms.filter((s) => s.kind === 'loopRef') as Array<{ name: string; scope: number }>;
        const decl = syms.find((s) => s.kind === 'loopDecl') as { scope: number };
        expect(refs.map((r) => r.name)).toEqual(['g']);
        expect(refs[0].scope).toBe(decl.scope);
      });

      it('a dotted loop reference marks only the root segment', () => {
        const src = '---\nparams { xs: pname[] }\n---\n{% for g in xs %}${g.field}{% endfor %}';
        const refs = symbols(src).filter((s) => s.kind === 'loopRef') as Array<{ name: string }>;
        expect(refs.map((r) => r.name)).toEqual(['g']);
      });

      it('lenient mode still collects loop symbols from a malformed template', () => {
        const bad = '---\nparams { xs: pname[] }\n---\n{% for g in xs %}${g} {% if';
        const kinds = symbols(bad).map((s) => s.kind);
        expect(kinds).toContain('loopDecl');
        expect(kinds).toContain('loopRef');
      });
    });
  });
});
